File: javascript/linkedlist/LinkedList.js
// Node构造函数
function Node(element) {
  this.element = element;
  this.prev = null;
  this.next = null;
}

function LinkedList() {
  this.head = new Node('head');
}

LinkedList.prototype.find = function(elem) {
  var pos = this.head;
  while (pos.next !== null) {
    pos = pos.next;
    if (pos.element === elem) return pos;
  }
  return null;
};

LinkedList.prototype.findLast = function() {
  var pos = this.head;
  while (pos.next !== null) {
    pos = pos.next;
  }
  return pos;
};

LinkedList.prototype.display = function() {
  var pos = this.head;
  while (pos.next !== null) {
    console.log(pos.next.element);
    pos = pos.next;
  }
};

LinkedList.prototype.dispReverse = function() {
  var pos = this.findLast();
  while (pos.prev !== null) {
    console.log(pos.prev.element);
    pos = pos.prev;
  }
};

LinkedList.prototype.insert = function(newElem) {
  var node = new Node(newElem);
  var last = this.findLast();
  if (last === null) {
    this.head.next = node;
    node.prev = this.head;
  }

  last.next = node;
  node.prev = last;
};

LinkedList.prototype.insertAfter = function(newElem, afterElem) {
  var node = new Node(newElem);
  var pos = this.find(afterElem);

  if (pos === null) return;
  if (pos.next !== null) {
    node.next = pos.next;
    pos.next.prev = node;
  }

  node.prev = pos;
  pos.next = node;
};

LinkedList.prototype.remove = function(elem) {
  var pos = this.find(elem);
  if (pos === null) return;

  if (pos.next !== null) {
    pos.next.prev = pos.prev;
  }
  pos.prev.next = pos.next;
  pos.prev = null;
  pos.next = null;
};


module.exports = LinkedList;

File: go/linkedlist/linkedlist.go
package linkedlist

type Lister interface {
	Clear()
	Len() int
	Front() *Element
	Back() *Element
	Remove(e *Element) interface{}
	PushFront(v interface{}) *Element
	PushBack(v interface{}) *Element
	InsertBefore(v interface{}, mark *Element) *Element
	InsertAfter(v interface{}, mark *Element) *Element
	MoveBefore(e, mark *Element)
	MoveAfter(e, mark *Element)
}

type Element struct {
	prev, next *Element
	list       *List
	Value      interface{}
}

func (e *Element) Next() *Element {
	if n := e.next; n != nil && n != &e.list.root {
		return n
	}
	return nil
}

func (e *Element) Prev() *Element {
	if p := e.prev; p != nil && p != &e.list.root {
		return p
	}
	return nil
}

type List struct {
	root Element
	len  int
}

func New() *List {
	l := new(List)
	l.Clear()
	return l
}

func (l *List) Clear() {
	if firstElem := l.root.next; firstElem != nil && firstElem.list == l {
		firstElem.prev = nil
	}
	if lastElem := l.root.prev; lastElem != nil && lastElem.list == l {
		lastElem.next = nil
	}
	l.root.next = &l.root
	l.root.prev = &l.root
	l.len = 0
}

func (l *List) Len() int {
	return l.len
}

func (l *List) Front() *Element {
	if l.len == 0 {
		return nil
	}
	return l.root.next
}

func (l *List) Back() *Element {
	if l.len == 0 {
		return nil
	}
	return l.root.prev
}

func (l *List) remove(e *Element) *Element {
	e.prev.next = e.next
	e.next.prev = e.prev
	e.prev = nil
	e.next = nil
	e.list = nil
	l.len--
	return e
}

func (l *List) Remove(e *Element) interface{} {
	if e.list == l {
		l.remove(e)
	}
	return e.Value
}

func (l *List) insert(e, at *Element) *Element {
	n := at.next
	at.next = e
	e.prev = at
	e.next = n
	n.prev = e
	e.list = l
	l.len++
	return e
}

func (l *List) insertValue(v interface{}, at *Element) *Element {
	return l.insert(&Element{Value: v}, at)
}

func (l *List) PushFront(v interface{}) *Element {
	return l.insertValue(v, &l.root)
}

func (l *List) PushBack(v interface{}) *Element {
	return l.insertValue(v, l.root.prev)
}

func (l *List) InsertBefore(v interface{}, mark *Element) *Element {
	if mark.list != l {
		return nil
	}
	return l.insertValue(v, mark.prev)
}

func (l *List) InsertAfter(v interface{}, mark *Element) *Element {
	if mark.list != l {
		return nil
	}
	return l.insertValue(v, mark)
}

func (l *List) MoveBefore(e, mark *Element) {
	if e.list != l || mark.list != l || e == mark {
		return
	}
	l.insert(l.remove(e), mark.prev)
}

func (l *List) MoveAfter(e, mark *Element) {
	if e.list != l || mark.list != l || e == mark {
		return
	}
	l.insert(l.remove(e), mark)
}
